[MRG] Add update sidecar json functionality using template

adam2392 opened the pull request:
PR Description
--------------

Closes: #458 

Thanks to @a-hurst for starting it on his branch and I borrowed the code and added the corresponding test. It so far worked for one of these datasets I needed to override  abunch, but I'm sure there could be flaws in the code still.

### Summary
1. `update_sidecars` function for JSONS
2. refactored `_bids_validate` to a `conftest.py`
3. unit test for `update_sidecars`.
4. ~~WIP~~ example of using `update_sidecars` function. Please lmk what you think and if this is in the right direction?
5. `write_raw_bids` now adds to the sidecar json file a bunch of RECOMMENDED fields, so that way it is easier for the `update_sidecar_files()` to determine which file to update.
6. `BIDSPath.match()` now has a kwarg `return_json=False`

### TODO
1. ~~determine if `BIDSPath.match()` should match `.json` files or not?~~

2. ~~add all RECOMMENDED and OPTIONAL fields when writing BIDS dataset (e.g. if `InstitutionName` is not in the sidecar json, then using a template will be difficult to match that we need to write that field into the sidecar json and not some other arbitrary json file. The other option is to maintain a dictionary of fields per sidecar file in the config, so that way you know which file to write to when you pass in a template.~~

### Misc question:

1. How can we best use templates to update tsv files? e.g. for `channels.tsv` and `electrodes.tsv` files?
2. Should we "provide" a template for ppl to download?

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

hoechenberger commented on doc/whats_new.rst:
```diff
@@ -31,7 +31,8 @@ Detailed list of changes
 ^^^^^^^^^^^^
-xxx
+
+- Add :func:`mne_bids.update.update_sidecars` to allow updating sidecar json files with a template JSON by `Adam Li`_ and `Austin Hurst`_ (`#601 <https://github.com/mne-tools/mne-bids/pull/601>`_)
```
```suggestion
- Add :func:`mne_bids.update.update_sidecars` to allow updating sidecar JSON files with a template JSON by `Adam Li`_ and `Austin Hurst`_ (`#601 <https://github.com/mne-tools/mne-bids/pull/601>`_)
```

hoechenberger commented on mne_bids/path.py:
```diff
@@ -599,8 +599,7 @@ def match(self):
         # Only keep files (not directories), and omit the JSON sidecars.
-        fnames = [f.name for f in fnames
-                  if f.is_file() and f.suffix != '.json']
+        fnames = [f.name for f in fnames if f.is_file()]
```
We made a conscious decision to exclude JSON files here – I don't remember the reason though. But this needs to be discussed. Also the comment doesn't make sense now anymore

adam2392 replied:
I couldn't find any explicit reason from the PR awhile back: https://github.com/mne-tools/mne-bids/pull/507/

The only reason I can see for excluding json is if you want a list of the actual raw EEG/MEg/ieeg datasets, rather then their sidecars too?

adam2392 replied:
Any major reason to exclude jsons? Or should I add a kwarg to `match(include_json=False)`?

sappelhoff commented:
just for etiquette reasons, could you attribute a-hurst in the commits that you borrowed? That works using this syntax:

```text
commit msg bla bla


Co-authored-by: Richard Höchenberger <richard.hoechenberger@gmail.com>
```

with such a commit message, GitHub will pick up, that this commit was done by you AND Richard (in this example)

see: https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors

for commits that already have a message, you could `git commit --amend` that commit

hoechenberger commented on mne_bids/tests/conftest.py:
```diff
@@ -0,0 +1,33 @@
+
+
+# WINDOWS issues:
+# the bids-validator development version does not work properly on Windows as
+# of 2019-06-25 --> https://github.com/bids-standard/bids-validator/issues/790
```
Actually the solution is right there in the issue: simply run the validator via `noda $executable`. This should also work on all platforms.

adam2392 replied:
I just copied this over from `test_write.py` into `conftest.py` to make it re-usable across multiple test files. Should I just remove these inline comments?

hoechenberger replied:
No leave the comments there so we will find them again in the future :)

hoechenberger commented on mne_bids/tests/test_update.py:
```diff
@@ -0,0 +1,108 @@
+"""Test for the MNE BIDS path functions."""
+# Authors: Adam Li <adam2392@gmail.com>
+#
+# License: BSD (3-clause)
```
I think the file name and the comment here are too generic

adam2392 replied:
I've updated it now. Lmk wyt.

hoechenberger commented on mne_bids/update.py:
```diff
@@ -0,0 +1,96 @@
+"""Update BIDS compatible directory structures and meta data."""
+# Authors: Adam Li <adam2392@gmail.com>
+#
+# License: BSD (3-clause)
+
+import json
```
The module name and the comment are too generic…

adam2392 replied:
I've updated it now. Lmk wyt.

hoechenberger commented on mne_bids/update.py:
```diff
@@ -0,0 +1,96 @@
+        _write_json(fpath, sidecar_json, overwrite=True, verbose=verbose)
+
+
+if __name__ == '__main__':
+    from pathlib import Path
+    from mne_bids import BIDSPath, get_entity_vals
+    WORKSTATION = "home"
+
+    if WORKSTATION == "home":
+        # bids root to write BIDS data to
+        bids_root = Path("/Users/adam2392/OneDrive - Johns Hopkins/sickkids/")
+        source_dir = bids_root / 'sourcedata'
+
+    elif WORKSTATION == "lab":
+        bids_root = Path("/home/adam2392/hdd2/epilepsy_bids/")
+        source_dir = Path("/home/adam2392/hdd2/epilepsy_bids/sourcedata")
+
+    subjects = get_entity_vals(bids_root, 'subject')
+    for subject in subjects:
+        bids_path = BIDSPath(subject=subject, datatype='ieeg', root=bids_root)
+        json_template = bids_root / 'sample_scripts' / 'json_template.json'
+        update_sidecars(bids_path, json_template)
```
This needs to go

adam2392 commented:
> just for etiquette reasons, could you attribute a-hurst in the commits that you borrowed? That works using this syntax:
> 
> ```
> commit msg bla bla
> 
> 
> Co-authored-by: Richard Höchenberger <richard.hoechenberger@gmail.com>
> ```
> 
> with such a commit message, GitHub will pick up, that this commit was done by you AND Richard (in this example)
> 
> see: https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors
> 
> for commits that already have a message, you could `git commit --amend` that commit

Oh okay didn't know this was possible!

So I just copy and pasted the necessary portions of the code from the branch he linked me to. I then did the commit ammend to the first commit in this branch. Does that work properly?

sappelhoff commented:
> jk Idk how to push up the commit. Tried this after git checkout 3bf09d0 and git commit --amend.

`git commit --amend` completely changes a commit --> the hash is different, and potentially the commit contents and/or the commit message. Thus, after using `git commit --amend`, you have re-written history and you have to use `git push --force` (like after a rebase).

I highly recommend to make a backup of your branch before doing the force push :-)

hoechenberger commented:
We need a good example, I honestly have no idea what is supposed to happen without actually running this code… :)

adam2392 commented:
> We need a good example, I honestly have no idea what is supposed to happen without actually running this code… :)

Kay can do! Should I append to an existing example, or add a whole new one altogether? Leaning towards new one so that way as we build out the "update" capability, then we can add there?

jasmainak commented:
> just for etiquette reasons, could you attribute a-hurst in the commits that you borrowed? That works using this syntax:

I normally cherry-pick the commit and then start from there :)

jasmainak commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,75 @@
+
+# TODO: add support for tsv files
+def update_sidecars(bids_path, sidecar_template_fpath, verbose=True):
+    """Update sidecar files using template JSON file.
```
I guess something that's not yet clear from documentation is that `bids_path` can match multiple paths. It should be clarified in the `BIDSPath` docstring and also here

adam2392 replied:
Sorry I'm not following. Isn't `BIDSPath.match()` intended to return multiple paths (i.e. a list of multiple paths that match the entities)?

jasmainak replied:
> Isn't BIDSPath.match() intended to return multiple paths

right, so an analogy is that `BIDSPath` is like a regular expression/wildcard not a `str`. It returns you the list of `str` when you use `match` (similar to `re.match`). The name makes `BIDSPath` makes you think it's a single file/directory and the docstring also says "Create a BIDS path." (note the use of singular). It's not clear up front that it could match more than one path. I'm not proposing to change the name, just that it should be clarified in docstrings

adam2392 replied:
Ah I see, yes that makes sense to do!

jasmainak commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,75 @@
+
+
+# TODO: add support for tsv files
+def update_sidecars(bids_path, sidecar_template_fpath, verbose=True):
```
I think it's simpler to provide a `dict` ... it's less of a hassle and everything is documented in your script -- hence more reproducible

hoechenberger replied:
Could be both, right? path-like or dict

adam2392 replied:
Included both.

hoechenberger commented:
> > We need a good example, I honestly have no idea what is supposed to happen without actually running this code… :)
> 
> 
> 
> Kay can do! Should I append to an existing example, or add a whole new one altogether? Leaning towards new one so that way as we build out the "update" capability, then we can add there?

Maybe start a new example and we can then see if it would make sense to integrate it into an existing one.

adam2392 commented:
> Maybe start a new example and we can then see if it would make sense to integrate it into an existing one.

Started new example and updated PR description with some things I think need to be addressed.

adam2392 commented:
Not sure why CI is failing since `make pep` and pytest worked locally. Will look into it later :/...

jasmainak commented on examples/update_bids_datasets.py:
```diff
@@ -0,0 +1,140 @@
+# For more information, you can check out the
+# respective :ref:`example <ex-convert-mne-sample>`.
+
+###############################################################################
```
is this line needed?

jasmainak commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,138 @@
+        # get the fields inside JSON
+        default_fields = set(sidecar_json.keys())
+
+        # XXX: how to we do this if fields are not in the sidecar file already?
```
I think it should work like a dictionary update in Python so folks can expect familiar behavior

jasmainak reviewed (commented):
Looks good so far. Can you make CircleCI happy so we can see the example?

Also this means that we make BIDS sidecars a first-class citizen and they would over-ride the metadata in the raw files. Can you verify (add tests?) if this is how the reading functions work? And also document this explicitly -- that in case of conflicts, the BIDS sidecar gets preference.

agramfort commented on examples/update_bids_datasets.py:
```diff
@@ -0,0 +1,140 @@
+import json
+from mne_bids import (BIDSPath, read_raw_bids,
+                      print_dir_tree, make_report)
+from mne_bids.sidecar_updates import update_sidecars
```
```suggestion
from mne_bids import update_sidecar
```

now this function should be quite straightforward:

```
for sc_fname in path.match():
     sc = json.load(sc_fname)
     sc.update(**values)
     json.dump(sc, sc_fname)
```

I don't get why such a small util function generates 400 lines of code.
If we are not careful of keeping things simple we are on a slippery slope.

hoechenberger replied:
> I don't get why such a small util function generates 400 lines of code.

To be fair, 1/4 of this is tests, and then there's also an example, so the **effective** number of new lines of production code is much smaller than 400

adam2392 commented:
> I don't get why such a small util function generates 400 lines of code.
> If we are not careful of keeping things simple we are on a slippery slope.

So the trouble comes up if we allow the API input to take a fully defined BIDSPath, or just a partially defined one. The function currently calls `bidspath.match()`, which will get ALL matching sidecar files according to the BIDSPath entities passed in. 

This then will have things like `coordsystem.json`, or `sidecar.json`, or maybe even `scans.json` (if you defined it yourself). If we just do:

```
sidecar_json.update(**template)
```

then it will add unwanted fields in a lot of files. Thoughts?

hoechenberger commented:
> So the trouble comes up if we allow the API input to take a fully defined BIDSPath, or just a partially defined one. The function currently calls `bidspath.match()`, which will get ALL matching sidecar files according to the BIDSPath entities passed in.

We should force users to also supply the suffix, no? Otherwise we'll have to be clever and guess, which leads to lots of lines of code.

Generally I believe we're trying to be too clever at many places of MNE-BIDS already, and would like us to reconsider certain design decisions from the past. But that's another story :)

adam2392 commented:
> We should force users to also supply the suffix, no? Otherwise we'll have to be clever and guess, which leads to lots of lines of code.

Okay then just to confirm, let's not use `match()`, but directly `BIDSPath.fpath` should exist and be a `.json` file rn and force users to define the full file path. This will actually make the code way easier.

adam2392 commented:
Any thoughts on how to do updating for .tsv files?

agramfort commented:
one thing at a time but please avoid doing "magic" to avoid passing some
things like suffix.

it will otherwise hit you back like a boomerang !

>

adam2392 commented:
Okay I've greatly simplified the `update_sidecar_json` function based on feedback. There's still some work to do to add unit tests and some additions to the example showing how to "update a bunch of json files" if desired.

adam2392 commented:
Not sure why whatsnew is acting funky. Anyways:

https://3405-89170358-gh.circle-artifacts.com/0/dev/auto_examples/update_bids_datasets.html#sphx-glr-auto-examples-update-bids-datasets-py

agramfort commented on doc/api.rst:
```diff
@@ -58,3 +58,19 @@ mne_bids.copyfiles
+.. autosummary::
+   :toctree: generated/
+
+   update_sidecar_json
```
don't add a full module for one function. Also 

from mne_bids.sidecar_updates import update_sidecar_json`

is very redundant. You repeat twice the same 2 words. I would do:

`from mne_bids import update_sidecar_json``

agramfort commented on examples/update_bids_datasets.py:
```diff
@@ -0,0 +1,147 @@
+"""
```
when an example does not start with plot_ it is not executed and rendered in the doc.

can you avoid this?

note that caching the mne somato data on circle ci is what mne does.

adam2392 replied:
Sorry I don't follow. Not as familiar with sphinx if that's the issue here. None of the other examples start with `plot_`, and the example is rendered here?

https://3407-89170358-gh.circle-artifacts.com/0/dev/auto_examples/update_bids_datasets.html#sphx-glr-auto-examples-update-bids-datasets-py

agramfort replied:
you're right ! see https://github.com/mne-tools/mne-bids/blob/master/doc/conf.py#L145

we execute every example even if it does not start with plot_

hoechenberger replied:
--> time to rename the examples in `mne-python` :)

jasmainak replied:
yeah this was a conscious choice because most of our examples don't "plot" anything.

agramfort commented on examples/update_bids_datasets.py:
```diff
@@ -0,0 +1,147 @@
+print(bids_path.match())
+
+# Define a sidecar template as a dictionary
+sidecar_template = {
```
can we call this maybe entries? I don't like the word template and sidecar is redundant

hoechenberger replied:
I agree that "template" may probably a little bit misleading

adam2392 replied:
Done. Changed to `entries`.

hoechenberger commented on examples/update_bids_datasets.py:
```diff
@@ -0,0 +1,147 @@
+"""
+========================
+11. Update BIDS datasets
```
```suggestion
12. Update BIDS datasets
```

adam2392 commented:
Not sure why windows is failing... but everything else LGTM.

Do we want to "encode" all the possible fields that each file can have as an error check, or simply assume user will be careful + bids-validator will get there one day? :p

agramfort commented on examples/update_bids_datasets.py:
```diff
@@ -0,0 +1,150 @@
+entries['Manufacturer'] = 'Elekta'
+entries['PowerLineFrequency'] = 50
+
+# create a temporary JSON file that contains the updated template
```
why do you need to do this? For an entry level example this looks like complex code. Can we avoid this?

hoechenberger replied:
If I'm not misunderstanding anything,`entries` can be a dict, so I agree we should pass a dict here. Passing a JSON file is advanced usage that maybe can be mentioned, but this example here is a bit too complicated

agramfort commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,170 @@
+
+
+# XXX: should we encode this, or let bids-validator do its job?
+# def _check_specification_compat(datatype):
```
if we go this route I would add the schema files to the mne-bids repo without any alteration.

eg https://github.com/bids-standard/bids-validator/blob/master/bids-validator/validators/json/schemas/meg.json

it means we need a python way to validate a json schema

hoechenberger replied:
> it means we need a python way to validate a json schema

https://python-jsonschema.readthedocs.io/en/stable/validate/

seems straightforward

jasmainak replied:
you're welcome: https://github.com/bids-standard/bids-validator/blob/master/bids-validator/bids_validator/bids_validator.py

hoechenberger replied:
> you're welcome: https://github.com/bids-standard/bids-validator/blob/master/bids-validator/bids_validator/bids_validator.py

But this is just for filename verification… not for JSON file content?

adam2392 replied:
Seems like ideally this is handled in pybids, or bids-validator. I'll remove these.

agramfort commented on mne_bids/path.py:
```diff
@@ -153,7 +153,8 @@ class BIDSPath(object):
     BIDSPath allows dynamic updating of its entities in place, and operates
-    similar to `pathlib.Path`.
+    similar to `pathlib.Path`. In addition, it can query multiple paths
+    with matching BIDS entities via the ``match`` function.
```
```suggestion
    with matching BIDS entities via the ``match`` method.
```

agramfort commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,134 @@
+    Will update metadata fields inside the path defined by
+    ``bids_path.fpath`` according to the ``entries``.
+    If a field does not exist in the corresponding sidecar file,
+    then that field will be created according to the template.
```
template?

can you do a pass on docstring @adam2392 ?

agramfort commented on mne_bids/tests/test_update.py:
```diff
@@ -0,0 +1,132 @@
+
+@pytest.fixture(scope='session')
+def sidecar_json_template(return_bids_test_dir):
+    """Return path to a sidecar JSON template."""
```
template?

agramfort commented on mne_bids/tests/test_update.py:
```diff
@@ -0,0 +1,132 @@
+
+
+@pytest.fixture(scope='session')
+def return_bids_test_dir(tmpdir_factory):
```
```suggestion
def _get_bids_test_dir(tmpdir_factory):
```

agramfort commented on mne_bids/tests/test_update.py:
```diff
@@ -0,0 +1,132 @@
+
+
+@pytest.fixture(scope='session')
+def sidecar_json_template(return_bids_test_dir):
```
```suggestion
def _get_sidecar_json(return_bids_test_dir):
```

agramfort commented on mne_bids/write.py:
```diff
@@ -536,6 +536,16 @@ def _sidecar_json(raw, task, manufacturer, fname, datatype, overwrite=False,
+        ('ManufacturersModelName', 'n/a'),
+        ('SoftwareVersions', 'n/a'),
+        ('TaskDescription', 'n/a'),
+    ]
```
is this necesary now?

agramfort commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,134 @@
+def update_sidecar_json(bids_path, entries, verbose=True):
+    """Update sidecar files using a dictionary or JSON file.
+
+    Will update metadata fields inside the path defined by
+    ``bids_path.fpath`` according to the ``entries``.
+    If a field does not exist in the corresponding sidecar file,
+    then that field will be created according to the ``entries``.
+    If a field does exist in the corresponding sidecar file,
+    then that field will be updated according to the ``entries``.
+
+    For example, if ``InstitutionName`` is
+    not defined in the sidecar json file, then trying to update
+    ``InstitutionName`` to ``Martinos Center`` will update
+    the sidecar json file to have ``InstitutionName`` as
+    ``Martinos Center``.
```
```suggestion
    Will update metadata fields inside the path defined by ``bids_path.fpath``
    according to the ``entries``. If a field does not exist in the
    corresponding sidecar file, then that field will be created according to
    the ``entries``. If a field does exist in the corresponding sidecar file,
    then that field will be updated according to the ``entries``.

    For example, if ``InstitutionName`` is not defined in the sidecar json
    file, then trying to update ``InstitutionName`` to ``Martinos Center``
    will update the sidecar json file to have ``InstitutionName`` as
    ``Martinos Center``.
```

agramfort commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,134 @@
+    the sidecar json file to have ``InstitutionName`` as
+    ``Martinos Center``.
+
+    Note that
```
cleanup

agramfort commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,134 @@
+        not set and only one data type (e.g., only EEG or MEG data)
+        is present in the dataset, it will be
+        selected automatically. This must uniquely identify
+        an existing file path, else an error will be raised.
```
please format close to 80 characters width

agramfort reviewed (commented):
some more feedback

adam2392 commented:
NOt sure why tests are failing when they succeed locally now...

hoechenberger commented:
@adam2392 Delete your testing data directory and run the tests again to have MNE download the data again. I once had a similar issue a while back and it was fixed through this…

hoechenberger commented:
Errm I mean, I had a similar inconsistency once. :)

adam2392 commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+    >>> # update sidecar coordsystem json file
+    >>> bids_path = BIDSPath(root='./', subject='001', session='001',
+                             suffix='coordsystem', extension='.json')
+    >>> entries = {'iEEGCoordinateSyste,': 'Other'}
```
```suggestion
    >>> entries = {'iEEGCoordinateSystem,': 'Other'}
```

hoechenberger commented:
@adam2392 Is this ready for another round of reviews? :)

adam2392 commented:
> @adam2392 Is this ready for another round of reviews? :)

Yes this is good to go now! Addressed all the feedback. Unless there's any more :p then can be merged?

Err I can fix conflicts first I see, but lmk if anymore changes to add too.

hoechenberger commented on doc/whats_new.rst:
```diff
@@ -27,15 +27,18 @@ Authors
 - The function :func:`mne_bids.print_dir_tree` has a new parameter ``return_str`` which allows it to return a str of the dir tree instead of printing it, by `Stefan Appelhoff`_ (`#600 <https://github.com/mne-tools/mne-bids/pull/600>`_)
 - :func:`mne_bids.write_raw_bids` now preserves event durations when writing :class:`mne.Annotations` to ``*_events.tsv`` files, and :func:`mne_bids.read_raw_bids` restores these durations upon reading, by `Richard Höchenberger`_ (`#603 <https://github.com/mne-tools/mne-bids/pull/603>`_)
 - Writing BrainVision data via :func:`mne_bids.write_raw_bids` will now set the unit of EEG channels to µV for enhanced interoperability with other software, by `Alexandre Gramfort`_, `Stefan Appelhoff`_, and `Richard Höchenberger`_ (`#610 <https://github.com/mne-tools/mne-bids/pull/610>`_)
+- Add :func:`mne_bids.update_sidecar_json` to allow updating sidecar JSON files with a template JSON by `Adam Li`_ and `Austin Hurst`_ (`#601 <https://github.com/mne-tools/mne-bids/pull/601>`_)
```
```suggestion
- Add :func:`mne_bids.update_sidecar_json` to allow updating sidecar JSON files with a template, by `Adam Li`_ and `Austin Hurst`_ (`#601 <https://github.com/mne-tools/mne-bids/pull/601>`_)
```

hoechenberger commented on examples/update_bids_datasets.py:
```diff
@@ -0,0 +1,137 @@
+    'InstitutionName': "Martinos Center"
+}
+
+# Note: ``update_sidecar_json`` will perform essentially a
+# dictionary update to your sidecar file, so be absolutely sure
+# that the ``entries`` are defined properly.
```
I'm not really sure what this is supposed to tell me?

adam2392 replied:
I've updated this description to just warn users that they want to be careful when updating since we don't do any checks based on valid fields per sidecar json file. It'll assume ``entries`` is defined correctly and properly for the corresponding `bids_path`.

hoechenberger commented on examples/update_bids_datasets.py:
```diff
@@ -0,0 +1,137 @@
+# Read the updated dataset
+# ------------------------
+#
+# We can use MNE-BIDS to update all sidecar files for a matching
+# ``BIDSPath`` object. We then update all matching metadata fields
+# within the BIDS dataset.
```
It seems this doesn't belong into this section..?

hoechenberger commented on mne_bids/path.py:
```diff
@@ -144,7 +144,7 @@ def _get_matched_empty_room(bids_path):
 
 class BIDSPath(object):
-    """Create a BIDS path.
+    """Create BIDS path object.
```
```suggestion
    """A BIDS path object.
```

hoechenberger commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+
+    Notes
+    -----
+    This function only works for JSON files.
```
```suggestion
    This function can only update JSON files.
```

hoechenberger commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+    You should double check that your update dictionary is correct
+    for the corresponding sidecar JSON file because it will perform
+    a dictionary update of the sidecar fields according to
+    the passed in dictionary overwriting any information that was
```
```suggestion
    the passed-in dictionary, overwriting any information that was
```

hoechenberger commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+    # get all matching json files
+    bids_path = bids_path.copy()
+    if bids_path.extension != '.json':
+        raise RuntimeError('Only works for ".json" files. The '
+                           'BIDSPath object passed in has '
+                           f'{bids_path.extension} extension.')
```
```suggestion
        raise RuntimeError(f'Only works for ".json" files. The '
                           f'BIDSPath object passed in has '
                           f'{bids_path.extension} extension.')
```

hoechenberger commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+    if isinstance(entries, dict):
+        sidecar_tmp = entries
+    else:
+        with open(entries, 'r') as tmp_f:
```
```suggestion
        with open(entries, 'r', encoding='utf-8-sig') as tmp_f:
```

hoechenberger commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+        print(f'Updating {fpath}...')
+
+    # load in sidecar filepath
+    with open(fpath, 'r') as tmp_f:
```
```suggestion
    with open(fpath, 'r', encoding='utf-8-sig') as tmp_f:
```

hoechenberger commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+            sidecar_tmp = json.load(
+                tmp_f, object_pairs_hook=OrderedDict)
+
+    if verbose:
+        print(sidecar_tmp)
+        print(f'Updating {fpath}...')
```
Use the logger here

hoechenberger commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+    val : str
+        The corresponding value to change to in the sidecar JSON file.
+    """
+    with open(sidecar_fname, "r") as fin:
```
```suggestion
    with open(sidecar_fname, 'r', encoding='utf-8-sig') as fin:
```

hoechenberger commented on mne_bids/sidecar_updates.py:
```diff
@@ -0,0 +1,136 @@
+    with open(sidecar_fname, "r") as fin:
+        sidecar_json = json.load(fin)
+    sidecar_json[key] = val
+    with open(sidecar_fname, "w") as fout:
```
```suggestion
    with open(sidecar_fname, 'w', encoding='utf-8') as fout:
```

hoechenberger commented on mne_bids/tests/test_update.py:
```diff
@@ -0,0 +1,131 @@
+    # get the sidecar json
+    sidecar_path = bids_path.copy().update(extension='.json')
+    sidecar_fpath = sidecar_path.fpath
+    with open(sidecar_fpath, 'r') as fin:
```
```suggestion
    with open(sidecar_fpath, 'r', encoding='utf-8') as fin:
```

hoechenberger commented on mne_bids/tests/test_update.py:
```diff
@@ -0,0 +1,131 @@
+
+    # update sidecars
+    update_sidecar_json(sidecar_path, _get_sidecar_json_update_file)
+    with open(sidecar_fpath, 'r') as fin:
```
```suggestion
    with open(sidecar_fpath, 'r', encoding='utf-8') as fin:
```

hoechenberger commented:
Thanks @adam2392!

agramfort commented on doc/whats_new.rst:
```diff
@@ -27,15 +27,22 @@ Authors
 
 Enhancements
 ^^^^^^^^^^^^
+<<<<<<< HEAD
```
@adam2392 careful next time :)

adam2392 replied:
Oh no! 

oops... Is this fixable with just a push into master by one of the devs?

sappelhoff replied:
one more note @adam2392 we now refer to PRs using the newly added sphinx substitution (#627):

```
(`#610 <https://github.com/mne-tools/mne-bids/pull/610>`_)
```

becomes
```
(:gh:`610`)
```

agramfort commented:
if you merge my PR https://github.com/mne-tools/mne-bids/pull/629 you will
be forgiven :)

>

hoechenberger commented:
Whoops:)

adam2392 commented:
> if you merge my PR #629 you will be forgiven :)
> […](#)

Can't merge it, but happily left a comment :).
